gapis/gfxapi/vulkan/vulkan_terminator.go

```diff
@@ -0,0 +1,378 @@
+			lrp = c.RenderPasses[t.RenderPass]
+			subpass = 0
+		case *RecreateCmdNextSubpassData:
+			subpass += 1
+		case *RecreateCmdEndRenderPassData:
+			lrp = nil
+			subpass = 0
+		}
+	}
+
+	walkCommands(c, queue.PendingCommands, f)
+	submitInfo := submit.PSubmits.Slice(uint64(0), uint64(submit.SubmitCount), l)
+	additional := 0
+	if len(idx) == 1 {
+		additional = 1
+	}
+	for sub := 0; sub < int(idx[0])+additional; sub++ {
+		info := submitInfo.Index(uint64(sub), l).Read(ctx, a, s, nil)
+		buffers := info.PCommandBuffers.Slice(uint64(0), uint64(info.CommandBufferCount), l)
+		for cmd := 0; cmd < int(info.CommandBufferCount); cmd++ {
+			buffer := buffers.Index(uint64(cmd), l).Read(ctx, a, s, nil)
+			bufferObject := c.CommandBuffers[buffer]
+			walkCommands(c, bufferObject.Commands, f)
+		}
+	}
+	if additional > 0 {
+		return lrp, subpass
+	}
+	if len(idx) == 2 {
+		additional = 1
+	}
+	lastInfo := submitInfo.Index(uint64(idx[0]), l).Read(ctx, a, s, nil)
+	lastBuffers := lastInfo.PCommandBuffers.Slice(uint64(0), uint64(lastInfo.CommandBufferCount), l)
+	for cmdbuffer := 0; cmdbuffer < int(idx[1])+additional; cmdbuffer++ {
+		buffer := lastBuffers.Index(uint64(cmdbuffer), l).Read(ctx, a, s, nil)
+		bufferObject := c.CommandBuffers[buffer]
+		walkCommands(c, bufferObject.Commands, f)
+	}
+	if additional > 0 {
+		return lrp, subpass
+	}
+	if len(idx) == 3 {
+		additional = 1
+	}
+	lastBuffer := lastBuffers.Index(uint64(idx[1]), l).Read(ctx, a, s, nil)
+	lastBufferObject := c.CommandBuffers[lastBuffer]
+	for cmd := 0; cmd < int(idx[2])+additional; cmd++ {
+		f(&lastBufferObject.Commands[cmd])
+	}
+	if additional > 0 {
+		return lrp, subpass
+	}
+	lastCommand := lastBufferObject.Commands[idx[2]]
+	if executeSubcommand, ok := (lastCommand).recreateData.(*RecreateCmdExecuteCommandsData); ok {
+		if len(idx) == 4 {
+			additional = 1
+		}
+		for subcmdidx := 0; subcmdidx < int(idx[3])+additional; subcmdidx++ {
+			buffer := executeSubcommand.CommandBuffers[uint32(subcmdidx)]
+			bufferObject := c.CommandBuffers[buffer]
+			walkCommands(c, bufferObject.Commands, f)
+		}
+		if additional > 0 {
+			return lrp, subpass
+		}
+		lastsubBuffer := executeSubcommand.CommandBuffers[uint32(idx[3])]
+		lastSubBufferObject := c.CommandBuffers[lastsubBuffer]
+		for subcmd := 0; subcmd < int(idx[4]); subcmd++ {
+			f(&lastSubBufferObject.Commands[subcmd])
+		}
+	}
+
+	return lrp, subpass
+}
+
+// RebuildCommandBuffer takes the commands from commandBuffer up to, and
+// including idx. It then appends any recreate* arguments to the end
+// of the command buffer.
+func RebuildCommandBuffer(ctx context.Context,
+	commandBuffer *CommandBufferObject,
+	s *gfxapi.State,
+	idx gfxapi.SubcommandIndex,
+	additionalCommands []interface{}) (VkCommandBuffer, []atom.Atom, []func()) {
+
+	x := make([]atom.Atom, 0)
+	cleanup := make([]func(), 0)
+
+	writeIdx := uint64(0)
+	if len(idx) > 0 {
+		writeIdx = idx[0]
+	}
+	additional := 0
+	if len(idx) == 1 {
+		additional = 1
+	}
```

Comment:
Hmm, this 'using writeIdx to determine if we are in a command buffer, and need an +1 when copying the original commands' is quite confusing to read. We probably needs a better way.

And as a new command buffer is always needed, we can creates that new command buffer first and that creation does not need `writeIdx` and `additional`, then put this 'copying the original commands, if any' logic in one place closer to each other.